torch/distributed/checkpoint/default_planner.py

```diff
@@ -72,11 +73,13 @@ def __init__(
         flatten_sharded_tensors: bool = True,
         dedup_replicated_tensors: Optional[bool] = None,
         dedup_save_to_lowest_rank: bool = False,
+        extensions: Optional[Sequence[StreamTransformExtension]] = None,
```

Comment:
I had kind of assumed this made more sense in the file writer since it's closely related to the serialization format, and generally we've thought about planners dealing with 'sharding' specific transforms. Am I missing something?

Reply:
Implementing the post processing as a buffer stream transformation makes it out of place in the Planner. I agree on that. Streaming, buffer reading/writing should all be internal to the storage IO. Planner should have no reason to be exposed to the internal mechanics of the storage components. 

However I see value in keeping Optimizer as another component which can be plugged into DCP. This gives the flexibility to the users in bringing the proprietary algorithms or at times our internal algorithms may not be open source, necessitating different versions of the optimizer hooks. 

Open question is where should these hooks be integrated in. If we take the approach of extending the buffer streams then it should be called in the storage components. I was more inclined to use the resolve hooks to do the post processing: sharding and format optimizations both. I prefer to keep the storage components as lean as possible, focusing only on the IO, serialization and deserialization only. Storage components are more likely to get extended/reimplemented by the partners as opposed to the planners. With IO, efficiency optimizations, storage resiliency optimizations, serialization/deserialization and also complex optimizations like compression, batching etc. this will grow significantly in complexity.
Planner should just get the data read from storage and then in resolve hooks, it should do whatever transformations are needed. 

Basically, the way I was thinking about this is that these storage optimizations should neither be in Planner nor in Storage reader/writer IMO but should be somewhat decoupled and invoked as general hooks by the planner at the resolve time.